### PR TITLE
[docs] deploying static ray cluster to K8S with external Redis for fault tolerance

### LINKED ISSUE #757: [Feature] [Docs] Docs on how to deploy on K8s without KubeRay
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Strictly speaking, this is not a KubeRay issue. However, I think it makes sense to track anything related to Ray on Kubernetes in the KubeRay repo.

Many users do not have the cluster-scoped permissions required to install the KubeRay CRDs in their K8s cluster. Other users prefer not to use custom controllers for natural reasons of risk aversion -- Kubernetes built-ins are generally more stable and reliable than community-maintained custom controllers.

For this reason, there should be a page of docs explaining how to manage a Ray cluster by manually creating deployments and services. 
There are actually docs on this in older releases of the Ray repo:
https://github.com/ray-project/ray/blob/releases/1.0.0/doc/source/cluster/kubernetes.rst

Should probably mention the downsides of not using KubeRay:
- No access to bells and whistles (autoscaling, fault-tolerant services, simplified config interfaces, etc.)
- More manual work required.
- KubeRay is the official solution, so other methods will not be as well supported.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by YQ-Wang**
## Why are these changes needed?
This PR adds the documentation and sample config files for deploying Ray to K8S without using KubeRay. As KubeRay CRDs need cluster-scoped permissions, this PR helps those users who do not have cluster-scoped permissions to install Ray Cluster in their K8S.

## Related issue number
Closes [#757](https://github.com/ray-project/kuberay/issues/757)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
For posterity: Earlier comments on this PR are at https://github.com/ray-project/ray/pull/31403

The doc build has an error here: https://buildkite.com/ray-project/oss-ci-build-pr/builds/10404#0185ed06-d921-4790-b29f-7d156b925b11/1697-2103

I think you can fix it by adding the new page here: https://github.com/ray-project/ray/blob/e910e67231ed81f5b1ed28dc4758743ac704398e/doc/source/_toc.yml#L304

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (kevin85421) [inline doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md:13]: Maybe we can add some instructions about Kubernetes cluster creation (e.g. `kind create cluster`) to enable users without Kubernetes knowledge to reproduce this doc.

**Review by kevin85421 — COMMENTED** [review]
Thank you for your contribution! I will test this document manually tomorrow.

**Comment by YQ-Wang** [comment]
> For posterity: Earlier comments on this PR are at #31403
> 
> The doc build has an error here: https://buildkite.com/ray-project/oss-ci-build-pr/builds/10404#0185ed06-d921-4790-b29f-7d156b925b11/1697-2103
> 
> I think you can fix it by adding the new page here:
> 
> https://github.com/ray-project/ray/blob/e910e67231ed81f5b1ed28dc4758743ac704398e/doc/source/_toc.yml#L304

@architkulkarni  Done, added the new doc to the toc yaml file.

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md:45]: Done, added a preparation section.

**Review by kevin85421 — COMMENTED** [review]
I cloned your branch and tried the following commands on my Ubuntu devbox:

```sh
kind create cluster # kindest/node:v1.24.0

# Deploy a sample Ray Cluster with fault tolerance enabled.
kubectl apply -f doc/source/cluster/kubernetes/configs/static-ray-cluster.with-fault-tolerance.yaml
```

The head Pod will keep crashing and I attach the error message in the following figure. Do I miss any step?

Screen Shot 2023-01-27 at 9 07 07 AM

Screen Shot 2023-01-27 at 9 05 22 AM

**Comment by YQ-Wang** [comment]
> I cloned your branch and tried the following commands on my Ubuntu devbox:
> 
> ```shell
> kind create cluster # kindest/node:v1.24.0
> 
> # Deploy a sample Ray Cluster with fault tolerance enabled.
> kubectl apply -f doc/source/cluster/kubernetes/configs/static-ray-cluster.with-fault-tolerance.yaml
> ```
> 
> The head Pod will keep crashing and I attach the error message in the following figure. Do I miss any step?
> 
> Screen Shot 2023-01-27 at 9 07 07 AM
> 
> Screen Shot 2023-01-27 at 9 05 22 AM

I tried yesterday as well and it works end to end. You need to have an external Redis service in your K8S and set the Redis ip:port(6379) in the static-ray-cluster.with-fault-tolerance.yaml (line 195).

**Comment by YQ-Wang** [comment]
> > @kevin85421 I tried yesterday as well and it works end to end. You need to have an external Redis service in your K8S and set the Redis ip:port(6379) in the static-ray-cluster.with-fault-tolerance.yaml (line 195).
> 
> Ah thanks! I guess setting up external redis on K8s is something every user of this doc will need to do to be successful. Would it make sense to add a sentence or two about how to do this? (Could even just be a link to an external guide.)

@architkulkarni Done, I have added a section in the config yaml and updated the doc. Now there is no additional work needed, should be able to just deploy by applying those two yamls.

**Comment by architkulkarni** [comment]
That's awesome! I just tested on my laptop and everything worked.

**Review by architkulkarni — APPROVED** [review]

**Comment by architkulkarni** [comment]
Windows and docker failures unrelated (this PR only touches documentation.)

The linkcheck failure " 404 Client Error: Not Found for url: https://raw.githubusercontent.com/ray-project/ray/master/doc/source/cluster/kubernetes/configs/static-ray-cluster.with-fault-tolerance.yaml" should go away once this PR is merged to master. (It's a chicken and egg situation)

**Comment by architkulkarni** [comment]
@ray-project/ray-docs can you please give codeowner approval?

**Comment by pcmoritz** [comment]
Yes, let me have a quick scan at the PR and then I'll approve it :)

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (pcmoritz) [inline doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md:123]: crasehs -> crashes

**Review by pcmoritz — APPROVED** [review]
Very nice, thanks a ton for you contribution! This is a great and simple way to deploy Ray on K8s for use cases that don't need more advanced features like autoscaling 👍

**Comment by pcmoritz** [comment]
Going forward, I think we can re-brand this as `manual` setup vs. `advanced` setup -- in many ways this is simpler than KubeRay -- but we should leave it as `advanced` until it is more battle tested :)

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md]: Done

**Review by kevin85421 — COMMENTED** [review]
Everything worked on my devbox!

In addition, I tried to test the functionality of fault tolerance by the following steps:
* Step 1: Install a detached actor in the Ray's namespace "detached_actor_ns"
* Step 2: Kill the GCS server process in the head Pod, and wait the head container to restart.
   `kubectl exec -it ${HEAD_POD} -- pkill gcs_server`
* Step 3: The detached actor needs tens of seconds to restore after the head Pod restarts.

Step 3 can access the detached actor, so the fault-tolerance feature works as expected. 

# One question about fault tolerance
[Ray v2 Architecture (Failure Model)](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview#heading=h.mduznt6qyr8a) mentions that "Ray worker nodes are designed to be homogeneous, so that any single node may be lost without bringing down the entire cluster. The current exception to this is the head node, since it hosts the GCS. In 2.0, we have added experimental support for GCS fault tolerance, which allows the GCS to be restarted while minimizing disturbance to the rest of the cluster.".

Hence, I expected that the entire cluster will be brought down after the `gcs_server` process is killed in Step 2. However, worker Pods were not impacted by the restart of the head Pod.

Maybe this fulfills the description "allows the GCS to be restarted while minimizing disturbance to the rest of the cluster."?

@architkulkarni do you know who is the expert of fault tolerance? Thank you!

**Comment by YQ-Wang** [comment]
> 



> > @architkulkarni do you know who is the expert of fault tolerance? Thank you!
> > I believe @iycheng is the expert here, but if I'm wrong he can redirect us.
> 
> I think the fault tolerance discussion won't block this PR -- @YQ-Wang once you decide how you want to address these last nits and comments, I will merge it.

@architkulkarni Done, thanks

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md]: Done

**Comment by kevin85421** [comment]
The fault tolerance discussion may cause the update of some Kubernetes configurations, including the liveness probe, readiness probe, and restart policy, to make them compatible with the fault-tolerance feature.

It's OK to merge this PR. We can submit a new patch if we find any issues.

**Review by kevin85421 — APPROVED** [review]

**Comment by architkulkarni** [comment]
> Windows and docker failures unrelated (this PR only touches documentation.)
> 
> The linkcheck failure " 404 Client Error: Not Found for url: https://raw.githubusercontent.com/ray-project/ray/master/doc/source/cluster/kubernetes/configs/static-ray-cluster.with-fault-tolerance.yaml" should go away once this PR is merged to master. (It's a chicken and egg situation)

Same as before, tests are okay. Merging

**Pull request merged**
